internal/datastore/has_perm.go

```diff
@@ -0,0 +1,111 @@
+package datastore
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+)
+
+const (
+	groupDefaultPK = 1
+	groupAdminPK   = 2
+)
+
+type hasPerm struct {
+	mu        sync.RWMutex
+	groupPerm map[int]map[string]bool
+	userGroup map[int][]int
+}
+
+func (h *hasPerm) update(data map[string]json.RawMessage) error {
+	h.mu.RLock()
+	defer h.mu.RUnlock()
+	for k, v := range data {
+		parts := strings.Split(k, ":")
+		if len(parts) != 2 {
+			return fmt.Errorf("key %s has wrong format. Expected one `:`", k)
+		}
+
+		var err error
+		switch parts[0] {
+		case "users/user":
+			err = h.updateUserGroup(v)
+		case "users/group":
+			err = h.updateGroupPerm(v)
+		}
+		if err != nil {
+			return fmt.Errorf("update %s: %w", k, err)
+		}
+	}
+	return nil
+}
+
+func (h *hasPerm) HasPerm(uid int, perm string) bool {
+	if uid == 0 {
+		return h.groupPerm[groupDefaultPK][perm]
+	}
+
+	for _, groupID := range h.userGroup[uid] {
+		if groupID == groupAdminPK {
+			// User is in admin group.
+			return true
+		}
+
+		return h.groupPerm[groupID][perm]
```

Comment:
Are you sure that this is correct? if the user has multiple groups, just the first one is checked.

Reply:
Thx for the review.

I fixed and tested it: https://github.com/OpenSlides/openslides3-autoupdate-service/pull/12

With the help of the tests, I found a second bug in this small piece of code...